mm/iob/iob_alloc.c

```diff
@@ -87,6 +87,15 @@ static FAR struct iob_s *iob_alloc_committed(void)
 
       g_iob_committed = iob->io_flink;
 
+      g_iob_sem.semcount--;
+
+#if CONFIG_IOB_THROTTLE > 0
+      if (g_throttle_sem.semcount > 0)
```

Comment:
- Question 1: How could `g_throttle_sem.semcount` be greater than 0 here?
- Question 2: I want to make sure you've considered these situations, which may relate to these codes:
  - Situation 1: `iob_alloc_committed` for **throttled** IOBs
    - Starting with `g_iob_sem.semcount = CONFIG_IOB_THROTTLE` and `g_throttle_sem.semcount = 0`
    - Task A: `iob_allocwait` with `throttled=true`, so `g_throttle_sem` is decreased by `nxsem_wait_uninterruptible` (now `g_throttle_sem.semcount = -1`)
    - Task B: `iob_free`, which posts to both `g_iob_sem` and `g_throttle_sem` (now `g_iob_sem.semcount = CONFIG_IOB_THROTTLE + 1` and `g_throttle_sem.semcount = 0`)
    - So I think we don't need to decrease `g_throttle_sem` here under this situation (yes we don't, because `g_throttle_sem.semcount` won't be greater than 0 here, and maybe never?), but we need to decrease `g_iob_sem`.
  - Situation 2: `iob_alloc_committed` for **non-throttled** IOBs
    - Starting with `g_iob_sem.semcount = 0`
    - Task A: `iob_allocwait` with `throttled=false`, so `g_iob_sem` is decreased by `nxsem_wait_uninterruptible` (now `g_iob_sem.semcount = -1`)
    - Task B: `iob_free`, which posts only to `g_iob_sem` (now `g_iob_sem.semcount = 0`)
    - Then shall we still decrease the sem on L90 under this case?

Reply:
Hi @wengzhe 

I'm sorry for the late response. There were some implementation flaws in these corner cases. In fact, we just need to decrement `g_iob_sem.semcount` when we have added the just-freed IOB to the `g_iob_committed` list **if it's going to be used by a task waiting for a throttled IOB**. Otherwise, we don't decrement `g_iob_sem.semcount` because it already represents the amount of free resources (in this case, less or equal to zero, if a task is waiting for a non-throttled IOB).

I simplified the implementation one step further. If an IOB is being freed and if a task is waiting for it, it will be put in the `g_iob_committed` list. If it is going to be committed to a task waiting for a throttled IOB, `committed_thottled` is set.  If set, we decrement it from the `g_iob_sem.semcount` before posting to the `g_throttle_sem`. No decrement occurs in `iob_allocwait`.

This satisfies the corner cases. The same tests were performed under similar results with this new implementation.